sidecar/register/healthchecker.go

```diff
@@ -71,34 +72,35 @@ func (checker *HealthChecker) Stop() {
 	checker.stop <- struct{}{}
 }
 
-func (checker *HealthChecker) report() {
+// maintainRegistration
+func (checker *HealthChecker) maintainRegistration() {
 	wasHealthy := false // Initialize as unhealthy so that we register on the first successful health check
 
-	statusChan := make(chan HealthStatus, len(checker.checks))
+	statusChan := make(chan healthcheck.Status, len(checker.checks))
 	for _, healthCheck := range checker.checks {
 		healthCheck.Start(statusChan)
 	}
 
 	// Set of health checks that have most recently reported unhealthy statuses.
-	unhealthyStatuses := make(map[HealthCheck]interface{})
+	unhealthyStatuses := make(map[healthcheck.Check]interface{})
```

Comment:
Having `healthcheck.Check` as the map key is problematic, because map keys must be [comparable](https://golang.org/ref/spec#Comparison_operators). Particularly, `healthcheck.Command` is not comparable because it has a slice field (`Args []string`) which isn't comparable, so this would have panicked at runtime.

We can make sure that each healthcheck is comparable, but that imposes some unjustified restrictions on the healthchecks.

Other solution would be to dismiss the `unhealthyStatuses` set in favor of a simple counter of how many healthchecks are currently healthy. This requires each healthcheck to have a channel of its own. Something along this line:

```go
healthyChan := make(chan bool, len(checker.checks))

for _, healthCheck := range checker.checks {
	statusChan := make(chan healthcheck.Status, 1)
	healthCheck.Start(statusChan)
	
	go func() {
		wasHealthy := false
		for {
			select {
			case status <- statusChan:
				healthy := status.Error == nil
				if healthy != wasHealthy {
					healthyChan <- healthy
				}
				wasHealthy = healthy
			}
		}
	}()
}

numHealthy := 0
for {
	select {
	case healthy <- healthyChan:
		if healthy {
			numHealthy++
			if numHealthy == len(checker.checks) {
				checker.registration.Start()
			}
		} else {
			numHealthy--
			if numHealthy == len(checker.checks) - 1 {
				checker.registration.Stop()
			}
		}
	}
}
```